wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/SystemManager.java

```diff
@@ -0,0 +1,115 @@
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.eclipse.che.commons.lang.concurrent.ThreadLocalPropagateContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.eclipse.che.api.system.shared.SystemStatus.PREPARING_TO_SHUTDOWN;
+import static org.eclipse.che.api.system.shared.SystemStatus.READY_TO_SHUTDOWN;
+import static org.eclipse.che.api.system.shared.SystemStatus.RUNNING;
+
+/**
+ * Facade for system operations.
+ *
+ * @author Yevhenii Voevodin
+ */
+@Singleton
+public class SystemManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SystemManager.class);
+
+    private final AtomicReference<SystemStatus> statusRef;
+    private final WorkspaceManager              workspaceManager;
+    private final EventService                  eventService;
+
+    private final CountDownLatch shutdownLatch = new CountDownLatch(1);
+
+    @Inject
+    public SystemManager(WorkspaceManager workspaceManager, EventService eventService) {
+        this.workspaceManager = workspaceManager;
+        this.eventService = eventService;
+        this.statusRef = new AtomicReference<>(RUNNING);
+    }
+
+    /**
+     * Stops some of the system services preparing system to lighter shutdown.
+     * System status is changed from {@link SystemStatus#RUNNING} to
+     * {@link SystemStatus#PREPARING_TO_SHUTDOWN}.
+     *
+     * @throws ConflictException
+     *         when system status is different from running
+     */
+    public void stopServices() throws ConflictException {
+        if (!statusRef.compareAndSet(RUNNING, PREPARING_TO_SHUTDOWN)) {
+            throw new ConflictException("System shutdown has been already called, system status: " + statusRef.get());
+        }
+        ExecutorService exec = Executors.newSingleThreadExecutor(
+                new ThreadFactoryBuilder().setDaemon(false)
+                                          .setNameFormat("ShutdownSystemServicesPool")
+                                          .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                                          .build());
+        exec.execute(ThreadLocalPropagateContext.wrap(this::doStopServices));
+        exec.shutdown();
+    }
+
+    /**
+     * Gets current system status.
+     *
+     * @see SystemStatus
+     */
+    public SystemStatus getSystemStatus() {
+        return statusRef.get();
+    }
+
+    /** Synchronously stops corresponding services. */
+    private void doStopServices() {
+        LOG.info("Preparing system to shutdown");
+        eventService.publish(new SystemStatusChangedEvent(RUNNING, PREPARING_TO_SHUTDOWN));
```

Comment:
I think there is an error on status/prevStatus
```java
 SystemStatusChangedEvent(SystemStatus status, SystemStatus prevStatus) {
```
so it means prev status was PREPARING_TO_SHUTDOWN and new status is RUNNING while IMHO it's the opposite